File: VB/E3597/VirtualMasterDetailTree.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraTreeList
Imports System.Data

Namespace E3597
    Friend Class VirtualMasterDetailTree
        Private _dataSource As DataTable
        Public Property DataSource() As DataTable
            Get
                Return _dataSource
            End Get
            Set(ByVal value As DataTable)
                If _dataSource Is value Then
                    Return
                End If
                _dataSource = value
                _masterTreeList.RefreshDataSource()
            End Set
        End Property
        Private _masterTreeList As treelist
        Private Property MasterTreeList() As treelist
            Get
                Return _masterTreeList
            End Get
            Set(ByVal value As treelist)
                If _masterTreeList Is value Then
                    Return
                End If
                Detach()
                Attach(value)
            End Set
        End Property
        Private RootObject As Object
        Public Sub New(ByVal masterTreeList As treelist, ByVal dt As DataTable)
            RootObject = New Object()
            Me.MasterTreeList = masterTreeList
            _dataSource = dt
        End Sub

        Private Sub Detach()
            If _masterTreeList Is Nothing Then
                Return
            End If
            _masterTreeList.DataSource = Nothing
            RemoveHandler _masterTreeList.VirtualTreeGetCellValue, AddressOf VirtualTreeGetCellValue
            RemoveHandler _masterTreeList.VirtualTreeGetChildNodes, AddressOf VirtualTreeGetChildNodes
            RemoveHandler _masterTreeList.VirtualTreeSetCellValue, AddressOf VirtualTreeSetCellValue

        End Sub
        Private Sub Attach(ByVal treeList As treelist)
            If treeList Is Nothing Then
                Return
            End If
            _masterTreeList = treeList
            _masterTreeList.DataSource = RootObject
            AddHandler _masterTreeList.VirtualTreeGetCellValue, AddressOf VirtualTreeGetCellValue
            AddHandler _masterTreeList.VirtualTreeGetChildNodes, AddressOf VirtualTreeGetChildNodes
            AddHandler _masterTreeList.VirtualTreeSetCellValue, AddressOf VirtualTreeSetCellValue
        End Sub

        Private Sub VirtualTreeGetChildNodes(ByVal sender As Object, ByVal e As VirtualTreeGetChildNodesInfo)
            If e.Node Is RootObject Then
                e.Children = _dataSource.Select()
            ElseIf TypeOf e.Node Is DataRow Then
                Dim row As DataRow = TryCast(e.Node, DataRow)
                If row.Table.ChildRelations.Count = 0 Then
                    Return
                End If


                e.Children = (TryCast(e.Node, DataRow)).GetChildRows(row.Table.ChildRelations(0))
            End If
        End Sub

        Private Sub VirtualTreeGetCellValue(ByVal sender As Object, ByVal e As VirtualTreeGetCellValueInfo)
            If (TryCast(e.Node, DataRow)).Table.Columns(e.Column.FieldName) IsNot Nothing Then
                Dim value As Object = (TryCast(e.Node, DataRow))(e.Column.FieldName)
                If value IsNot Nothing Then
                    e.CellData = value
                End If
            End If
        End Sub

        Private Sub VirtualTreeSetCellValue(ByVal sender As Object, ByVal e As VirtualTreeSetCellValueInfo)
            If (TryCast(e.Node, DataRow)).Table.Columns(e.Column.FieldName) IsNot Nothing Then
                TryCast(e.Node, DataRow)(e.Column.FieldName) = e.NewCellData
            End If
        End Sub

    End Class
End Namespace

File: CS/E3597/VirtualMasterDetailTree.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraTreeList;
using System.Data;

namespace treelist
{
    class VirtualMasterDetailTree
    {
        DataTable _dataSource;
        public DataTable DataSource
        {
            get { return _dataSource; }
            set
            {
                if (_dataSource == value)
                    return;
                _dataSource = value;
                _masterTreeList.RefreshDataSource();
            }
        }
        TreeList _masterTreeList;
        TreeList MasterTreeList
        {
            get { return _masterTreeList; }
            set
            {
                if (_masterTreeList == value)
                    return;
                Detach();
                Attach(value);
            }
        }
        object RootObject;
        public VirtualMasterDetailTree(TreeList masterTreeList, DataTable dt)
        {
            RootObject = new object();
            MasterTreeList = masterTreeList;
            _dataSource = dt;
        }

        void Detach()
        {
            if (_masterTreeList == null)
                return;
            _masterTreeList.DataSource = null;
            _masterTreeList.VirtualTreeGetCellValue -= new VirtualTreeGetCellValueEventHandler(VirtualTreeGetCellValue);
            _masterTreeList.VirtualTreeGetChildNodes -= new VirtualTreeGetChildNodesEventHandler(VirtualTreeGetChildNodes);
            _masterTreeList.VirtualTreeSetCellValue -= new VirtualTreeSetCellValueEventHandler(VirtualTreeSetCellValue);

        }
        void Attach(TreeList treeList)
        {
            if (treeList == null)
                return;
            _masterTreeList = treeList;
            _masterTreeList.DataSource = RootObject;
            _masterTreeList.VirtualTreeGetCellValue +=new VirtualTreeGetCellValueEventHandler(VirtualTreeGetCellValue);
            _masterTreeList.VirtualTreeGetChildNodes += new VirtualTreeGetChildNodesEventHandler(VirtualTreeGetChildNodes);
            _masterTreeList.VirtualTreeSetCellValue+=new VirtualTreeSetCellValueEventHandler(VirtualTreeSetCellValue);
        }

        private void VirtualTreeGetChildNodes(object sender, VirtualTreeGetChildNodesInfo e)
        {
            if (e.Node == RootObject)
                e.Children = _dataSource.Select();
            else if (e.Node is DataRow)
            {
                DataRow row = e.Node as DataRow;
                if (row.Table.ChildRelations.Count == 0) return;


                e.Children = (e.Node as DataRow).GetChildRows(row.Table.ChildRelations[0]);
            }
        }

        private void VirtualTreeGetCellValue(object sender, VirtualTreeGetCellValueInfo e)
        {
            if ((e.Node as DataRow).Table.Columns[e.Column.FieldName] != null)
            {
                object value = (e.Node as DataRow)[e.Column.FieldName];
                if (value != null)
                    e.CellData = value;
            }
        }

        private void VirtualTreeSetCellValue(object sender, VirtualTreeSetCellValueInfo e)
        {
            if ((e.Node as DataRow).Table.Columns[e.Column.FieldName] != null)
            {
                (e.Node as DataRow)[e.Column.FieldName] = e.NewCellData;
            }
        }

    }
}
